nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/main/java/org/apache/nifi/processors/gcp/bigquery/PutBigQueryBatch.java

```diff
@@ -0,0 +1,263 @@
+        return new PropertyDescriptor.Builder()
+                .name(propertyDescriptorName)
+                .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+                .expressionLanguageSupported(true)
+                .dynamic(true)
+                .build();
+    }
+
+    @Override
+    @OnScheduled
+    public void onScheduled(ProcessContext context) {
+        super.onScheduled(context);
+
+        if (schemaCache == null) {
+            String schemaStr = context.getProperty(TABLE_SCHEMA).evaluateAttributeExpressions().getValue();
+            schemaCache = BqUtils.schemaFromString(schemaStr);
+        }
+    }
+
+    @Override
+    public void onTrigger(ProcessContext context, ProcessSession session) throws ProcessException {
+        FlowFile flowFile = session.get();
+        if (flowFile == null) {
+            return;
+        }
+
+        final Map<String, String> attributes = new HashMap<>();
+
+        final BigQuery bq = getCloudService();
+
+        final String projectId = context.getProperty(PROJECT_ID).evaluateAttributeExpressions().getValue();
+        final String dataset = context.getProperty(DATASET).evaluateAttributeExpressions(flowFile).getValue();
+        final String tableName = context.getProperty(TABLE_NAME).evaluateAttributeExpressions(flowFile).getValue();
+
+        final TableId tableId;
+        if (StringUtils.isEmpty(projectId)) {
+            tableId = TableId.of(dataset, tableName);
+        } else {
+            tableId = TableId.of(projectId, dataset, tableName);
+        }
+
+        final String fileType = context.getProperty(SOURCE_TYPE).getValue();
+
+        WriteChannelConfiguration writeChannelConfiguration =
+                WriteChannelConfiguration.newBuilder(tableId)
+                        .setCreateDisposition(JobInfo.CreateDisposition.valueOf(context.getProperty(CREATE_DISPOSITION).getValue()))
+                        .setWriteDisposition(JobInfo.WriteDisposition.valueOf(context.getProperty(WRITE_DISPOSITION).getValue()))
+                        .setIgnoreUnknownValues(context.getProperty(IGNORE_UNKNOWN).asBoolean())
+                        .setMaxBadRecords(context.getProperty(MAXBAD_RECORDS).asInteger())
+                        .setSchema(schemaCache)
+                        .setFormatOptions(FormatOptions.of(fileType))
+                        .build();
+        TableDataWriteChannel writer = bq.writer(writeChannelConfiguration);
+
+        try {
+            session.read(flowFile, rawIn -> {
+                ReadableByteChannel readableByteChannel = Channels.newChannel(rawIn);
+                ByteBuffer byteBuffer = ByteBuffer.allocateDirect(BUFFER_SIZE);
+                while (readableByteChannel.read(byteBuffer) > 0) {
+                    byteBuffer.flip();
+                    writer.write(byteBuffer);
+                    byteBuffer.clear();
+                }
+            });
```

Comment:
Does anyone have any thoughts on the above? It seems when I run from an integration test, it works great. However when I actually use it in NiFi, it writes 0 byte tables. Any thoughts? I moved to the above to help work with large files (where the file is too large to fit in a byte array).

Reply:
@zenfenan if you have any input on the above, that would be great. Thanks in advance!